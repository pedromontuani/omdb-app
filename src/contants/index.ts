import { IFilterOption } from '../components/filter-modal';

export const OMDB_API_KEY = '6cefdc45';

export const FILTER_OPTIONS: IFilterOption[] = [
  { value: 'ALPHABETICAL', label: 'Alphabetical order' },
  { value: 'RATING', label: 'Rating' },
];
