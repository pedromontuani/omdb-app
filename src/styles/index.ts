import colors from './colors';
import measures from './measures';

const theme = {
  colors,
  measures,
};

export default theme;
