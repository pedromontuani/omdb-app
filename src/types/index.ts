export interface IFilm {
  Poster: string;
  Title: string;
  Type: string;
  Year: string;
  imdbID: string;
  details: IFilmDetails;
}

export interface IRating {
  Source: 'Internet Movie Database' | 'Rotten Tomatoes' | 'Metacritic';
  Value: string;
}

export type IFilmDetails = {
  Actors: string;
  Awards: string;
  BoxOffice: string;
  Country: string;
  DVD: string;
  Director: string;
  Genre: string;
  Language: string;
  Metascore: string;
  Plot: string;
  Poster: string;
  Production: string;
  Rated: string;
  Ratings: IRating[];
  Released: string;
  Response: string;
  Runtime: string;
  Title: string;
  Type: string;
  Website: string;
  Writer: string;
  Year: string;
  imdbID: string;
  imdbRating: string;
  imdbVotes: string;
  avgRating: number;
};

export interface IFilmsQueryResponse {
  Search: IFilm[];
  Response: 'True' | 'False';
  Error?: string;
  totalResults: number;
}
