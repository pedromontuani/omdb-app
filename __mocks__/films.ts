import { IFilm } from '../src/types';

export const MOCK_FILM: IFilm = {
  Poster: 'https://picsum.photos/200',
  Title: 'Mock title',
  Type: 'Mock type',
  Year: '2020',
  imdbID: 'mock imdbID',
  details: {
    Actors: 'Mock actors',
    Awards: 'Mock awards',
    BoxOffice: 'Mock box office',
    Country: 'Mock Country',
    DVD: 'Mock DVD',
    Director: 'Mock Director',
    Genre: 'Mock Genre',
    Language: 'Mock Language',
    Metascore: 'Mock Metascore',
    Plot: 'Mock Plot',
    Poster: 'https://picsum.photos/200',
    Production: 'Mock Production',
    Rated: 'Mock Rated',
    Ratings: [
      {
        Source: 'Internet Movie Database',
        Value: '7.5/10',
      },
    ],
    Released: 'Mock Released',
    Response: 'Mock Response',
    Runtime: 'Mock Runtime',
    Title: 'Mock Title',
    Type: 'Mock Type',
    Website: 'Mock Website',
    Writer: 'Mock Writer',
    Year: 'Mock Year',
    imdbID: 'Mock imdbID',
    imdbRating: 'Mock imdbRating',
    imdbVotes: '7.5',
    avgRating: 7.5,
  },
};

export const MOCK_FILMS_LIST: IFilm[] = [
  MOCK_FILM,
  MOCK_FILM,
  MOCK_FILM,
  MOCK_FILM,
  MOCK_FILM,
  MOCK_FILM,
].map((film, index) => ({
  ...film,
  imdbID: index.toString(),
  details: { ...film.details, imdbID: index.toString() },
}));
