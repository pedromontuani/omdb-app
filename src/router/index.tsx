import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';

import HomeScreen from '../screens/home';
import { IFilmDetails } from '../types';
import DetailsScreen from '../screens/details';

export type RootStackParamList = {
  Home: undefined;
  Details: {
    filmDetails: IFilmDetails;
  };
};

const Stack = createNativeStackNavigator<RootStackParamList>();

const Router: React.FC<{}> = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator
        screenOptions={{
          headerShown: false,
        }}>
        <Stack.Screen name="Home" component={HomeScreen} />
        <Stack.Screen name="Details" component={DetailsScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default Router;
