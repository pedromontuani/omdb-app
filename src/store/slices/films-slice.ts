import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IFilm } from '../../types';

export interface IFilmsState {
  films: Array<any>;
  sort: 'ALPHABETICAL' | 'RATING';
  year?: string;
  query: string;
  status: 'success' | 'pending' | 'error' | null;
  error: string | null;
}

export const initialState: IFilmsState = {
  films: [],
  sort: 'ALPHABETICAL',
  query: '',
  year: undefined,
  status: null,
  error: null,
};

const filmsSlice = createSlice({
  name: 'films',
  initialState,
  reducers: {
    GET_FILMS: (state, _: PayloadAction<IGetFilmsAction>) => ({
      ...state,
      films: [],
      status: 'pending',
    }),
    GET_FILMS_SUCCESS: (
      state,
      { payload }: PayloadAction<IGetFilmesSuccessAction>,
    ) => ({
      ...state,
      ...payload,
      status: 'success',
    }),
    GET_FILMS_ERROR: (state, { payload }: PayloadAction<IErrorAction>) => ({
      ...state,
      ...payload,
      status: 'error',
    }),
    SET_SORT: (state, { payload }: PayloadAction<ISetSortAction>) => ({
      ...state,
      sort: payload.sort,
    }),
  },
});

interface IErrorAction {
  error: string;
}

interface IGetFilmsAction {
  query: string;
  year?: string;
}

interface IGetFilmesSuccessAction {
  films: IFilm[];
  query: string;
}

interface ISetSortAction {
  sort: 'ALPHABETICAL' | 'RATING';
}

const { actions, reducer } = filmsSlice;

export { reducer };

export const { GET_FILMS, GET_FILMS_ERROR, GET_FILMS_SUCCESS, SET_SORT } =
  actions;
