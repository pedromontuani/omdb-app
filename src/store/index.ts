import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';
import reduxFlipper from 'redux-flipper';
import { TypedUseSelectorHook, useSelector, useDispatch } from 'react-redux';
import { all } from 'redux-saga/effects';

import { reducer as filmsReducer } from './slices/films-slice';

import filmsSaga from './sagas/films-saga';

const sagaMiddleware = createSagaMiddleware();

const middlewares = [sagaMiddleware, ...(__DEV__ ? [reduxFlipper()] : [])];

function* rootSaga() {
  yield all([filmsSaga()]);
}

export const store = configureStore({
  reducer: {
    films: filmsReducer,
  },
  middleware: middlewares,
});

sagaMiddleware.run(rootSaga);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAppDispatch = () => useDispatch<AppDispatch>();
