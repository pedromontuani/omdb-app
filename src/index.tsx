import React from 'react';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { Provider } from 'react-redux';
import styled, { ThemeProvider } from 'styled-components/native';

import Router from './router';
import theme from './styles';
import { store } from './store';
import { ToastProvider } from './context/toast-context';

const App: React.FC<{}> = () => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <SafeAreaProvider>
          <ToastProvider>
            <StyledStatusBar barStyle="light-content" />
            <Router />
          </ToastProvider>
        </SafeAreaProvider>
      </ThemeProvider>
    </Provider>
  );
};

const StyledStatusBar = styled.StatusBar.attrs(({ theme: { colors } }) => ({
  backgroundColor: colors.primaryDark,
}))``;

export default App;
