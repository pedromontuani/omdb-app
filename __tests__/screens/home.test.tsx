import React from 'react';
import HomeScreen from '../../src/screens/home';

import { customRender, rendererCreate } from '../utils';
import { initialState } from '../../src/store/slices/films-slice';
import * as filmsSlice from '../../src/store/slices/films-slice';
import * as store from '../../src/store';
import { MOCK_FILMS_LIST } from '../../__mocks__/films';
import { fireEvent } from '@testing-library/react-native';

jest.useFakeTimers();

describe('testing home screen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  }),
    it('should match snapshot', async () => {
      const tree = rendererCreate(<HomeScreen />);
      expect(tree).toMatchSnapshot();
    });

  it('should render no-data component', async () => {
    const { getByTestId } = customRender(<HomeScreen />);

    const noDataComponent = await getByTestId('no-data');

    expect(noDataComponent).toBeTruthy();
  });

  it('should render the films cards', async () => {
    const useSelector = jest.spyOn(store, 'useAppSelector');
    useSelector.mockImplementationOnce(selectorFn =>
      selectorFn({ films: { ...initialState, films: MOCK_FILMS_LIST } }),
    );

    const { getAllByTestId } = customRender(<HomeScreen />);

    const cards = await getAllByTestId('film-card');

    expect(cards).toHaveLength(MOCK_FILMS_LIST.length);
  });

  it('should call search action', async () => {
    const MOCK_QUERY = 'Avengers';
    const GET_FILMS = jest.spyOn(filmsSlice, 'GET_FILMS');
    const useSelector = jest.spyOn(store, 'useAppSelector');
    useSelector.mockImplementationOnce(selectorFn =>
      selectorFn({ films: { ...initialState, films: MOCK_FILMS_LIST } }),
    );

    const { getByTestId } = customRender(<HomeScreen />);

    const searchInput = await getByTestId('search-input');
    const searchButton = await getByTestId('search-button');

    await fireEvent.changeText(searchInput, MOCK_QUERY);
    await fireEvent.press(searchButton);

    expect(GET_FILMS).toHaveBeenCalledWith({ query: MOCK_QUERY });
  });
});
