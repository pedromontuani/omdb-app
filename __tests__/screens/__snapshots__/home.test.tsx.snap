// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing home screen should match snapshot 1`] = `
Array [
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "backgroundColor": "#eee",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#f1f2f3",
            "elevation": 5,
            "flexDirection": "row",
            "height": 60,
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 0,
            "shadowColor": "rgba(0, 0, 0, 0.25)",
            "shadowOffset": Object {
              "height": 1,
              "width": 2,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3.8,
            "zIndex": 100,
          },
        ]
      }
    >
      <TextInput
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search for films by name..."
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
        testID="search-input"
        value=""
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "marginRight": -20,
            "opacity": 1,
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 0,
          }
        }
        testID="search-button"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#bbb",
                  "fontSize": 24,
                },
              ],
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <RCTScrollView
      ListEmptyComponent={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": [Function],
        }
      }
      contentContainerStyle={
        Object {
          "padding": 20,
        }
      }
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingTop": 80,
              },
            ]
          }
          testID="no-data"
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/no-data/no-data.png",
              }
            }
          />
          <Text
            style={
              Array [
                Object {
                  "color": "#7f7e94",
                  "fontSize": 16,
                  "marginTop": 60,
                },
              ]
            }
          >
            Please use the search bar
          </Text>
        </View>
      </View>
    </RCTScrollView>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#5F49AB",
          "borderRadius": 30,
          "bottom": 20,
          "elevation": 6,
          "height": 60,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "right": 20,
          "width": 60,
          "zIndex": 100,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 28,
              },
            ],
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
    <Modal
      animationType="fade"
      hardwareAccelerated={true}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    />
  </RNCSafeAreaView>,
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#dc3545",
        "borderRadius": 8,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 40,
        "position": "absolute",
        "transform": Array [
          Object {
            "translateY": -120,
          },
        ],
        "width": 718,
        "zIndex": 100,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
            "marginLeft": 16,
            "marginRight": 16,
            "marginTop": 16,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 22,
            },
            Array [
              Object {},
            ],
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "marginLeft": 16,
              "marginRight": 16,
            },
          ]
        }
      />
    </View>
  </View>,
]
`;
